File: Software/StarPRNT SDK/ObjectiveC SDK/ObjectiveC SDK/ModelCapability.m
//
//  ModelCapability.m
//  ObjectiveC SDK
//
//  Created by Yuji on 2015/**/**.
//  Copyright © 2015年 Star Micronics. All rights reserved.
//

#import "ModelCapability.h"

typedef NS_ENUM(NSInteger, ModelCapabilityIndex) {
    ModelCapabilityIndexTitle = 0,
    ModelCapabilityIndexEmulation,
    ModelCapabilityIndexCashDrawerOpenActive,
    ModelCapabilityIndexPortSettings,
    ModelCapabilityIndexModelNameArray
};

static const NSArray      *_modelIndexArray;
static const NSDictionary *_modelCapabilityDictionary;

@implementation ModelCapability

+ (void)initialize {
    if (self == [ModelCapability class]) {
        _modelIndexArray = @ [
            @(ModelIndexMPOP),
            @(ModelIndexFVP10),
            @(ModelIndexTSP100),
            @(ModelIndexTSP650II),
            @(ModelIndexTSP700II),
            @(ModelIndexTSP800II),
            @(ModelIndexSP700),                 // <-
            @(ModelIndexSM_S210I),
            @(ModelIndexSM_S220I),
            @(ModelIndexSM_S230I),
            @(ModelIndexSM_T300I),
            @(ModelIndexSM_T400I),
            @(ModelIndexSM_L200),               // <-
            @(ModelIndexBSC10),
            @(ModelIndexSM_S210I_StarPRNT),
            @(ModelIndexSM_S220I_StarPRNT),
            @(ModelIndexSM_S230I_StarPRNT),
            @(ModelIndexSM_T300I_StarPRNT),
            @(ModelIndexSM_T400I_StarPRNT)
//          @(ModelIndexSM_L200),
//          @(ModelIndexSP700)
        ];
        
        _modelCapabilityDictionary = @ {
            @(ModelIndexMPOP)              : @[@"mPOP",              @(StarIoExtEmulationStarPRNT),      @NO,  @"",         @[@"POP10"]],
            @(ModelIndexFVP10)             : @[@"FVP10",             @(StarIoExtEmulationStarLine),      @YES, @"",         @[@"FVP10 (STR_T-001)"]],      // Only LAN model
            @(ModelIndexTSP100)            : @[@"TSP100",            @(StarIoExtEmulationStarGraphic),   @YES, @"",         @[@"TSP113",
                                                                                                                              @"TSP143"]],
            @(ModelIndexTSP650II)          : @[@"TSP650II",          @(StarIoExtEmulationStarLine),      @YES, @"",         @[@"TSP654II (STR_T-001)",     // Only LAN model->
                                                                                                                              @"TSP654 (STR_T-001)",
                                                                                                                              @"TSP651 (STR_T-001)"]],
            @(ModelIndexTSP700II)          : @[@"TSP700II",          @(StarIoExtEmulationStarLine),      @YES, @"",         @[@"TSP743II (STR_T-001)",
                                                                                                                              @"TSP743 (STR_T-001)"]],
            @(ModelIndexTSP800II)          : @[@"TSP800II",          @(StarIoExtEmulationStarLine),      @YES, @"",         @[@"TSP847II (STR_T-001)",
                                                                                                                              @"TSP847 (STR_T-001)"]],     // <-Only LAN model
            @(ModelIndexSM_S210I)          : @[@"SM-S210i",          @(StarIoExtEmulationEscPosMobile),  @NO,  @"mini",     @[@"SM-S210i"]],               // Sample->
            @(ModelIndexSM_S220I)          : @[@"SM-S220i",          @(StarIoExtEmulationEscPosMobile),  @NO,  @"mini",     @[@"SM-S220i"]],
            @(ModelIndexSM_S230I)          : @[@"SM-S230i",          @(StarIoExtEmulationEscPosMobile),  @NO,  @"mini",     @[@"SM-S230i"]],
            @(ModelIndexSM_T300I)          : @[@"SM-T300i",          @(StarIoExtEmulationEscPosMobile),  @NO,  @"mini",     @[@"SM-T300i"]],
            @(ModelIndexSM_T400I)          : @[@"SM-T400i",          @(StarIoExtEmulationEscPosMobile),  @NO,  @"mini",     @[@"SM-T400i"]],               // <-Sample
            @(ModelIndexBSC10)             : @[@"BSC10",             @(StarIoExtEmulationEscPos),        @YES, @"escpos",   @[@"BSC10"]],
            @(ModelIndexSM_S210I_StarPRNT) : @[@"SM-S210i StarPRNT", @(StarIoExtEmulationStarPRNT),      @NO,  @"Portable", @[@"SM-S210i StarPRNT"]],      // Sample->
            @(ModelIndexSM_S220I_StarPRNT) : @[@"SM-S220i StarPRNT", @(StarIoExtEmulationStarPRNT),      @NO,  @"Portable", @[@"SM-S220i StarPRNT"]],
            @(ModelIndexSM_S230I_StarPRNT) : @[@"SM-S230i StarPRNT", @(StarIoExtEmulationStarPRNT),      @NO,  @"Portable", @[@"SM-S230i StarPRNT"]],
            @(ModelIndexSM_T300I_StarPRNT) : @[@"SM-T300i StarPRNT", @(StarIoExtEmulationStarPRNT),      @NO,  @"Portable", @[@"SM-T300i StarPRNT"]],
            @(ModelIndexSM_T400I_StarPRNT) : @[@"SM-T400i StarPRNT", @(StarIoExtEmulationStarPRNT),      @NO,  @"Portable", @[@"SM-T400i StarPRNT"]],      // <-Sample
            @(ModelIndexSM_L200)           : @[@"SM-L200",           @(StarIoExtEmulationStarPRNT),      @NO,  @"Portable", @[@"SM-L200"]],
            @(ModelIndexSP700)             : @[@"SP700",             @(StarIoExtEmulationStarDotImpact), @YES, @"",         @[@"SP712 (STR-001)",          //Only LAN model
                                                                                                                              @"SP717 (STR-001)",
                                                                                                                              @"SP742 (STR-001)",
                                                                                                                              @"SP747 (STR-001)"]]
        };
    }
}

+ (NSInteger)modelIndexCount {
    return _modelIndexArray.count;
}

+ (ModelIndex)modelIndexAtIndex:(NSInteger)index {
    return [_modelIndexArray[index] integerValue];
}

+ (NSString *)titleAtModelIndex:(ModelIndex)modelIndex {
    return _modelCapabilityDictionary[@(modelIndex)][ModelCapabilityIndexTitle];
}

+ (StarIoExtEmulation)emulationAtModelIndex:(ModelIndex)modelIndex {
    return [_modelCapabilityDictionary[@(modelIndex)][ModelCapabilityIndexEmulation] integerValue];
}

+ (BOOL)cashDrawerOpenActiveAtModelIndex:(ModelIndex)modelIndex {
    return [_modelCapabilityDictionary[@(modelIndex)][ModelCapabilityIndexCashDrawerOpenActive] boolValue];
}

+ (NSString *)portSettingsAtModelIndex:(ModelIndex)modelIndex {
    return _modelCapabilityDictionary[@(modelIndex)][ModelCapabilityIndexPortSettings];
}

+ (ModelIndex)modelIndexAtModelName:(NSString *)modelName {
    for (id modelIndex in _modelCapabilityDictionary) {
        NSArray *modelNameArray = _modelCapabilityDictionary[modelIndex][ModelCapabilityIndexModelNameArray];
        
        for (int i = 0; i < modelNameArray.count; i++) {
            if ([modelName hasPrefix:modelNameArray[i]] == YES) {
                return [modelIndex integerValue];
            }
        }
    }
    
    return ModelIndexNone;
}

@end

File: Software/StarPRNT SDK/Swift SDK/Swift SDK/ModelCapability.swift
//
//  ModelCapability.swift
//  Swift SDK
//
//  Created by Yuji on 2015/**/**.
//  Copyright © 2015年 Star Micronics. All rights reserved.
//

import Foundation

enum ModelIndex: Int {     // Don't insert(Only addition)
    case None = 0
    case MPOP
    case FVP10
    case TSP100
    case TSP650II
    case TSP700II
    case TSP800II
    case SM_S210I
    case SM_S220I
    case SM_S230I
    case SM_T300I
    case SM_T400I
    case BSC10
    case SM_S210I_StarPRNT
    case SM_S220I_StarPRNT
    case SM_S230I_StarPRNT
    case SM_T300I_StarPRNT
    case SM_T400I_StarPRNT
    case SM_L200
    case SP700
}

class ModelCapability : NSObject {
    enum ModelCapabilityIndex: Int {
        case Title = 0
        case Emulation
        case CashDrawerOpenActive
        case PortSettings
        case ModelNameArray
    }
    
    static let modelIndexArray: [ModelIndex] = [
        ModelIndex.MPOP,
        ModelIndex.FVP10,
        ModelIndex.TSP100,
        ModelIndex.TSP650II,
        ModelIndex.TSP700II,
        ModelIndex.TSP800II,
        ModelIndex.SP700,                 // <-
        ModelIndex.SM_S210I,
        ModelIndex.SM_S220I,
        ModelIndex.SM_S230I,
        ModelIndex.SM_T300I,
        ModelIndex.SM_T400I,
        ModelIndex.SM_L200,               // <-
        ModelIndex.BSC10,
        ModelIndex.SM_S210I_StarPRNT,
        ModelIndex.SM_S220I_StarPRNT,
        ModelIndex.SM_S230I_StarPRNT,
        ModelIndex.SM_T300I_StarPRNT,
        ModelIndex.SM_T400I_StarPRNT
//      ModelIndex.SM_L200,
//      ModelIndex.SP700
    ]
    
    static var modelCapabilityDictionary: [ModelIndex: [AnyObject]] = [
        ModelIndex.MPOP              : ["mPOP",              StarIoExtEmulation.StarPRNT     .rawValue, false, "",         ["POP10"]],
        ModelIndex.FVP10             : ["FVP10",             StarIoExtEmulation.StarLine     .rawValue, true,  "",         ["FVP10 (STR_T-001)"]],      // Only LAN model
        ModelIndex.TSP100            : ["TSP100",            StarIoExtEmulation.StarGraphic  .rawValue, true,  "",         ["TSP113",
                                                                                                                            "TSP143"]],
        ModelIndex.TSP650II          : ["TSP650II",          StarIoExtEmulation.StarLine     .rawValue, true,  "",         ["TSP654II (STR_T-001)",     // Only LAN model->
                                                                                                                            "TSP654 (STR_T-001)",
                                                                                                                            "TSP651 (STR_T-001)"]],
        ModelIndex.TSP700II          : ["TSP700II",          StarIoExtEmulation.StarLine     .rawValue, true,  "",         ["TSP743II (STR_T-001)",
                                                                                                                            "TSP743 (STR_T-001)"]],
        ModelIndex.TSP800II          : ["TSP800II",          StarIoExtEmulation.StarLine     .rawValue, true,  "",         ["TSP847II (STR_T-001)",
                                                                                                                            "TSP847 (STR_T-001)"]],     // <-Only LAN model
        ModelIndex.SM_S210I          : ["SM-S210i",          StarIoExtEmulation.EscPosMobile .rawValue, false, "mini",     ["SM-S210i"]],               // Sample->
        ModelIndex.SM_S220I          : ["SM-S220i",          StarIoExtEmulation.EscPosMobile .rawValue, false, "mini",     ["SM-S220i"]],
        ModelIndex.SM_S230I          : ["SM-S230i",          StarIoExtEmulation.EscPosMobile .rawValue, false, "mini",     ["SM-S230i"]],
        ModelIndex.SM_T300I          : ["SM-T300i",          StarIoExtEmulation.EscPosMobile .rawValue, false, "mini",     ["SM-T300i"]],
        ModelIndex.SM_T400I          : ["SM-T400i",          StarIoExtEmulation.EscPosMobile .rawValue, false, "mini",     ["SM-T400i"]],               // <-Sample
        ModelIndex.BSC10             : ["BSC10",             StarIoExtEmulation.EscPos       .rawValue, true,  "escpos",   ["BSC10"]],
        ModelIndex.SM_S210I_StarPRNT : ["SM-S210i StarPRNT", StarIoExtEmulation.StarPRNT     .rawValue, false, "Portable", ["SM-S210i StarPRNT"]],      // Sample->
        ModelIndex.SM_S220I_StarPRNT : ["SM-S220i StarPRNT", StarIoExtEmulation.StarPRNT     .rawValue, false, "Portable", ["SM-S220i StarPRNT"]],
        ModelIndex.SM_S230I_StarPRNT : ["SM-S230i StarPRNT", StarIoExtEmulation.StarPRNT     .rawValue, false, "Portable", ["SM-S230i StarPRNT"]],
        ModelIndex.SM_T300I_StarPRNT : ["SM-T300i StarPRNT", StarIoExtEmulation.StarPRNT     .rawValue, false, "Portable", ["SM-T300i StarPRNT"]],
        ModelIndex.SM_T400I_StarPRNT : ["SM-T400i StarPRNT", StarIoExtEmulation.StarPRNT     .rawValue, false, "Portable", ["SM-T400i StarPRNT"]],      // <-Sample
        ModelIndex.SM_L200           : ["SM-L200",           StarIoExtEmulation.StarPRNT     .rawValue, false, "Portable", ["SM-L200"]],
        ModelIndex.SP700             : ["SP700",             StarIoExtEmulation.StarDotImpact.rawValue, true,  "",         ["SP712 (STR-001)",          //Only LAN model
                                                                                                                            "SP717 (STR-001)",
                                                                                                                            "SP742 (STR-001)",
                                                                                                                            "SP747 (STR-001)"]]
    ]
    
    static func modelIndexCount() -> Int {
        return ModelCapability.modelIndexArray.count
    }
    
    static func modelIndexAtIndex(index: Int) -> ModelIndex {
        return ModelCapability.modelIndexArray[index]
    }
    
    static func titleAtModelIndex(modelIndex: ModelIndex) -> String! {
        return ModelCapability.modelCapabilityDictionary[modelIndex]![ModelCapabilityIndex.Title.rawValue] as! String
    }
    
    static func emulationAtModelIndex(modelIndex: ModelIndex) -> StarIoExtEmulation {
        return StarIoExtEmulation(rawValue: ModelCapability.modelCapabilityDictionary[modelIndex]![ModelCapabilityIndex.Emulation.rawValue] as! Int)!
    }
    
    static func cashDrawerOpenActiveAtModelIndex(modelIndex: ModelIndex) -> Bool {
        return ModelCapability.modelCapabilityDictionary[modelIndex]![ModelCapabilityIndex.CashDrawerOpenActive.rawValue] as! Bool
    }
    
    static func portSettingsAtModelIndex(modelIndex: ModelIndex) -> String! {
        return ModelCapability.modelCapabilityDictionary[modelIndex]![ModelCapabilityIndex.PortSettings.rawValue] as! String
    }
    
    static func modelIndexAtModelName(modelName: String!) -> ModelIndex {
        for (modelIndex, anyObject) in ModelCapability.modelCapabilityDictionary {
            let modelNameArray: [String] = anyObject[ModelCapabilityIndex.ModelNameArray.rawValue] as! [String]
            
//          for var i: Int = 0; i < modelNameArray.count; i += 1 {
            for     i: Int in 0 ..< modelNameArray.count         {
                if modelName.hasPrefix(modelNameArray[i]) == true {
                    return modelIndex
                }
            }
        }
        
        return ModelIndex.None
    }
}
